contracts/src/base/registry/facets/distribution/v2/StakingRewards.sol

```diff
@@ -0,0 +1,490 @@
+    address stakeToken;
+    uint256 totalStaked;
+    uint256 rewardDuration;
+    uint256 rewardEndTime;
+    uint256 lastUpdateTime;
+    uint256 rewardRate;
+    uint256 rewardPerTokenAccumulated;
+    uint256 nextDepositId;
+    mapping(address depositor => uint256 amount) stakedByDepositor;
+    mapping(address beneficiary => Treasure) treasureByBeneficiary;
+    mapping(uint256 depositId => Deposit) depositById;
+    mapping(uint256 depositId => address proxy) proxyById;
+  }
+
+  event DelegationProxyDeployed(
+    uint256 depositId,
+    address delegatee,
+    address proxy
+  );
+
+  /*´:°•.°+.*•´.*:˚.°*.˚•´.°:°•.°•.*•´.*:˚.°*.˚•´.°:°•.°+.*•´.*:*/
+  /*                           ERRORS                           */
+  /*.•°:°.´+˚.*°.˚:*.´•*.+°.•°:´*.´•*.•°.•°:°.´:•˚°.*°.˚:*.´+°.•*/
+
+  error StakingRewards__InvalidAmount();
+  error StakingRewards__InvalidAddress();
+  error StakingRewards__InvalidRewardRate();
+  error StakingRewards__InsufficientReward();
+
+  /*´:°•.°+.*•´.*:˚.°*.˚•´.°:°•.°•.*•´.*:˚.°*.˚•´.°:°•.°+.*•´.*:*/
+  /*                          VIEWERS                           */
+  /*.•°:°.´+˚.*°.˚:*.´•*.+°.•°:´*.´•*.•°.•°:°.´:•˚°.*°.˚:*.´+°.•*/
+
+  function lastTimeRewardDistributed(
+    Layout storage ds
+  ) internal view returns (uint256) {
+    return FixedPointMathLib.min(ds.rewardEndTime, block.timestamp);
+  }
+
+  function currentRewardPerTokenAccumulated(
+    Layout storage ds
+  ) internal view returns (uint256) {
+    // cache storage reads
+    (
+      uint256 totalStaked,
+      uint256 lastUpdateTime,
+      uint256 rewardRate,
+      uint256 rewardPerTokenAccumulated
+    ) = (
+        ds.totalStaked,
+        ds.lastUpdateTime,
+        ds.rewardRate,
+        ds.rewardPerTokenAccumulated
+      );
+    if (totalStaked == 0) return rewardPerTokenAccumulated;
+
+    uint256 elapsedTime;
+    unchecked {
+      elapsedTime = lastTimeRewardDistributed(ds) - lastUpdateTime;
+    }
+    return
+      rewardPerTokenAccumulated +
+      FixedPointMathLib.fullMulDiv(rewardRate, elapsedTime, totalStaked);
+  }
+
+  function currentUnclaimedReward(
+    Layout storage ds,
+    Treasure storage treasure
+  ) internal view returns (uint256) {
+    uint256 rewardPerTokenGrowth;
+    unchecked {
+      rewardPerTokenGrowth =
+        currentRewardPerTokenAccumulated(ds) -
+        treasure.rewardPerTokenAccumulated;
+    }
+    return
+      treasure.unclaimedRewardSnapshot +
+      FixedPointMathLib.fullMulDiv(
+        treasure.earningPower,
+        rewardPerTokenGrowth,
+        SCALE_FACTOR
+      );
+  }
+
+  /*´:°•.°+.*•´.*:˚.°*.˚•´.°:°•.°•.*•´.*:˚.°*.˚•´.°:°•.°+.*•´.*:*/
+  /*                       STATE MUTATING                       */
+  /*.•°:°.´+˚.*°.˚:*.´•*.+°.•°:´*.´•*.•°.•°:°.´:•˚°.*°.˚:*.´+°.•*/
+
+  function updateGlobalReward(Layout storage ds) internal {
+    ds.rewardPerTokenAccumulated = currentRewardPerTokenAccumulated(ds);
+    ds.lastUpdateTime = lastTimeRewardDistributed(ds);
+  }
```

Comment:
missing event

Reply:
No implementation emits events for these as they are not useful offchain. It only increases gas.